lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,245 @@
+
+private:
+  PatchReadFirstLane(const PatchReadFirstLane &) = delete;
+  PatchReadFirstLane &operator=(const PatchReadFirstLane &) = delete;
+
+  bool liftReadFirstLane(Function &function);
+
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis.
+};
+
+} // namespace lgc
+
+// =====================================================================================================================
+// Initializes static members.
+char lgc::PatchReadFirstLane::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for readfirstlane optimizations.
+FunctionPass *lgc::createPatchReadFirstLane() {
+  return new PatchReadFirstLane();
+}
+
+PatchReadFirstLane::PatchReadFirstLane() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in,out] function : LLVM function to be run on.
+bool PatchReadFirstLane::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Read-First-Lane\n");
+
+  m_divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+
+  const bool changed = liftReadFirstLane(function);
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+}
+
+// =====================================================================================================================
+// Try to lift readfirstlanes per block to transform SGPR instructions into VGPR instruction.
+// Reversely iterate over a block who has readfirstlane instructions. Try to find qualified instructions to progagate
+// readfirstlane to them. The qualified instruction is that all its users are proven to be uniform and the cost keeps
+// one readfirstlane inserted for a lift.
+//
+// @param [in,out] function : LLVM function to be run for readfirstlane optimization.
+bool PatchReadFirstLane::liftReadFirstLane(Function &function) {
+  // Collect initial amdgcn_readfirstlane and their corresponding blocks
+  DenseSet<BasicBlock *> readFirstLaneBlocks;
+  DenseSet<Instruction *> initialReadFirstLanes;
+  Module *module = function.getParent();
+  for (auto &func : *module) {
+    if (func.getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      for (User *user : func.users()) {
+        Instruction *inst = cast<Instruction>(user);
+        if (inst->getFunction() != &function)
+          continue;
+        initialReadFirstLanes.insert(inst);
```

Comment:
Putting the initialReadFirstLanes that may belong to different blocks seems not a good idea. I think a map<BasicBlock*, DenseSet<Instruction *>> may help here, you can put the instructions that belong to the same block together.

Reply:
Got it. Done.